blog post (#2)
YosefLab/scvi-tools-site: blog/2021-3-03-scvi-tools-v090.md

---
slug: v090
title:      scvi-tools 0.9.0 release
date:       2021-03-03
author: Adam Gayoso, Romain Lopez, Galen Xing, Nir Yosef
tags: [scvi-tools, release]
---
import useBaseUrl from '@docusaurus/useBaseUrl';

Today we officially released `scvi-tools` version 0.9.0 ([changelog](https://docs.scvi-tools.org/en/stable/release_notes/index.html)). This release marks the culmination of five months of work on the backend of the codebase, which came after three months of work on the frontend.
In this short note, we officially introduce `scvi-tools` as a readily usable codebase that contains many implementations of probabilistic single-cell omics methods, and also features a high-level interface to accelerate the model development process. We start with some historical notes about our previous codebase, which was mostly used for internal developments in the last three years. We then describe the obstacles we found to its external adoption, and the foundational idea behind the new `scvi-tools` work: a high-level deep probabilistic programming library specialized for single-cell omics data.

<!--truncate-->

## Taking a step back: the original `scvi` codebase

> Many members of the Yosef Lab, and in particular Jeff Regier, Edouard Melhman, Romain, and Adam helped conceive, develop, and maintain the `scvi` codebase. The initial philosophy was to make the code available for users to run [scVI](https://www.nature.com/articles/s41592-018-0229-2), but also to have a proper codebase for developing novel algorithms for single-cell omics analysis (e.g., [totalVI](https://www.nature.com/articles/s41592-020-01050-x) was developed this way). Over the last three years, we have hosted seven visiting graduate students who wrote their master theses in the Yosef Lab by building new functionalities, as well as new algorithms (including [scANVI](https://www.embopress.org/doi/full/10.15252/msb.20209620), [AutoZI](https://www.biorxiv.org/content/biorxiv/early/2019/10/10/794875.full.pdf) and [gimVI](https://arxiv.org/pdf/1905.02269.pdf)) directly into the `scvi` codebase. However, our ambition at the time was to solely use this infrastructure for internal developments. Consequently, some mistakes were made, in which we did not make use of the existing infrastructure. For example, we had built manual data formatting boilerplate code to read many different single-cell omics input formats.

Author: ```suggestion
Many members of the Yosef Lab, and in particular Jeff Regier, Edouard Melhman, Romain, and Adam helped conceive, develop, and maintain the `scvi` codebase. The initial philosophy was to make the code available for users to run [scVI](https://www.nature.com/articles/s41592-018-0229-2), but also to have a proper codebase for developing novel algorithms for single-cell omics analysis (e.g., [totalVI](https://www.nature.com/articles/s41592-020-01050-x) was developed this way). Over the last three years, we have hosted seven visiting graduate students who wrote their master theses in the Yosef Lab by building new functionalities, as well as new algorithms (including [scANVI](https://www.embopress.org/doi/full/10.15252/msb.20209620), [AutoZI](https://www.biorxiv.org/content/biorxiv/early/2019/10/10/794875.full.pdf) and [gimVI](https://arxiv.org/pdf/1905.02269.pdf)) directly into the `scvi` codebase. However, our ambition at the time was to solely use this infrastructure for internal developments. Consequently, some mistakes were made, in which we did not make use of the existing single-cell analysis infrastructure. For example, we had built manual data formatting boilerplate code to read many different single-cell omics input formats.
```

---
slug: v090
title:      scvi-tools 0.9.0 release
date:       2021-03-03
author: Adam Gayoso, Romain Lopez, Galen Xing, Nir Yosef
tags: [scvi-tools, release]
---
import useBaseUrl from '@docusaurus/useBaseUrl';

Today we officially released `scvi-tools` version 0.9.0 ([changelog](https://docs.scvi-tools.org/en/stable/release_notes/index.html)). This release marks the culmination of five months of work on the backend of the codebase, which came after three months of work on the frontend.

> In this short note, we officially introduce `scvi-tools` as a readily usable codebase that contains many implementation of probabilistic single-cell omics methods, and also features a high-level interface to accelerate the model development process. We start with some historical notes about our previous codebase, which was mostly used for internal developments in the last three years. We then describe the obstacles we found to its external adoption, and the foundational idea behind the new `scvi-tools` work: a high-level deep probabilistic programming library specialized for single-cell omics data.

Reviewer: ```suggestion
In this short note, we officially introduce `scvi-tools` as a readily usable codebase that contains many implementations of probabilistic single-cell omics methods, and also features a high-level interface to accelerate the model development process. We start with some historical notes about our previous codebase, which was mostly used for internal developments in the last three years. We then describe the obstacles we found to its external adoption, and the foundational idea behind the new `scvi-tools` work: a high-level deep probabilistic programming library specialized for single-cell omics data.
```

<!--truncate-->

## Taking a step back: the original `scvi` codebase

> Many members of the Yosef Lab, and in particular Jeff Regier, Edouard Melhman, Romain and Adam helped conceive, develop and maintain the `scvi` codebase. The initial philosophy was to make available the code for users to run [scVI](https://www.nature.com/articles/s41592-018-0229-2), but also to have a proper codebase for developing novel algorithms for single-cell omics analysis (e.g., [totalVI](https://www.nature.com/articles/s41592-020-01050-x) was developed this way). Over the last three years, we hosted seven visiting graduate students who wrote their master theses in the Yosef Lab by building new functionalities, as well as new algorithms (including [scANVI](https://www.embopress.org/doi/full/10.15252/msb.20209620), [AutoZI](https://www.biorxiv.org/content/biorxiv/early/2019/10/10/794875.full.pdf) and [gimVI](https://arxiv.org/pdf/1905.02269.pdf)) directly into the `scvi` codebase. However, our ambition at the time was to solely use this infrastructure for internal developments. Consequently, some mistakes were made, in which we did not make use of existing infrastructure. For example, we had built manual data formatting boilerplate code to read many different single-cell omics input formats.

Reviewer: ```suggestion
Many members of the Yosef Lab, and in particular Jeff Regier, Edouard Melhman, Romain, and Adam helped conceive, develop, and maintain the `scvi` codebase. The initial philosophy was to make the code available for users to run [scVI](https://www.nature.com/articles/s41592-018-0229-2), but also to have a proper codebase for developing novel algorithms for single-cell omics analysis (e.g., [totalVI](https://www.nature.com/articles/s41592-020-01050-x) was developed this way). Over the last three years, we have hosted seven visiting graduate students who wrote their master theses in the Yosef Lab by building new functionalities, as well as new algorithms (including [scANVI](https://www.embopress.org/doi/full/10.15252/msb.20209620), [AutoZI](https://www.biorxiv.org/content/biorxiv/early/2019/10/10/794875.full.pdf) and [gimVI](https://arxiv.org/pdf/1905.02269.pdf)) directly into the `scvi` codebase. However, our ambition at the time was to solely use this infrastructure for internal developments. Consequently, some mistakes were made, in which we did not make use of the existing infrastructure. For example, we had built manual data formatting boilerplate code to read many different single-cell omics input formats.
```

Reviewer: I just added a few oxford commas

## New implementations

With these new model building blocks, we were able to implement models external to our lab with relative ease. This included [Stereoscope](https://www.nature.com/articles/s42003-020-01247-y) for deconvolving spatial transcriptomics data, [Solo](https://www.sciencedirect.com/science/article/pii/S2405471220301952) for detection of doublets in scRNA-seq data, and [CellAssign](https://www.nature.com/articles/s41592-019-0529-1) for reference-based annotation of scRNA-seq data. These implementations required significantly fewer lines of code with `scvi-tools`.

## Pyro integration

> Another feature we are excited about is integration with [Pyro](https://pyro.ai/), which further abstracts the process of manually deriving optimization objetives, thus allowing for complex hierarchical Bayesian models. Relying on Pyro to power a model is compeltely optional, but we feel there will be situations where it is clearly advantageous to use it.

Reviewer: ```suggestion
Another feature we are excited about is integration with [Pyro](https://pyro.ai/), which further abstracts the process of manually deriving optimization objectives, thus allowing for complex hierarchical Bayesian models. Relying on Pyro to power a model is completely optional, but we feel there will be situations where it is clearly advantageous to use it.
```
